snapshot-labs/snapshot
# Add UMA execution to SafeSnap

Fixes #2814 

**Changes in this PR:**
This pull request adds UMA execution as an option in SafeSnap. Our design segregates the UMA elements as much as possible from the existing Snapshot code, to make review easier and prevent the introduction of unexpected bugs.

In effect, the code should detect when the user has specified an UMA module address in their configuration, and detect a value tied to the UMA Optimistic Governor contract's `rules`. If the code sees that the user has an UMA module configured, they will be directed to a distinct `HandleOutcomeUma` view that uses the `umaModule` file.

There are also various configuration and other changes in the codebase, with a light touch, trying to leave Reality execution and everything else alone.

**How to test and review this PR?**
We've been testing functionality in a Vercel preview in our repo.
https://snapshot-uma.vercel.app/

The Outcome Test Snapshot Space is tied to a Safe with an UMA governance module and an UMA module address set in the configuration.
https://snapshot-uma.vercel.app/#/outcome.eth

I believe you can join the Space to create new proposals and test the end-to-end flow, but let me know if you run into any issues. The Space is set up on Goerli with bonds in Goerli ETH.

Edit: Looks like you have your own Vercel deployment flow but I don't have authorization to trigger it on this PR.

You should be able to successfully:

1. Create a proposal
2. Add transactions to the proposal
3. Vote on the proposal
4. Approve a bond
5. Request execution of a proposal (after the voting period)
6. Dispute execution through the testnet oracle UI (https://testnet.oracle.umaproject.org/)
7. Delete disputed proposals through the Snapshot UI
8. Propose again
9. Execute proposals after the challenge period if it is not disputed
10. See when a proposal has been completely executed
11. Create a different or duplicate proposal, and run through the same flow

If desired, we can also help you set up a new Safe, UMA module, and Snapshot Space to test the whole set-up from scratch. Until we have this merged, and a parallel PR merged with the Zodiac repo, the UMA module contracts will need to be deployed manually rather than through the Zodiac app.

In the end state, the user set-up flow will look like:
`Create Safe -> Create Snapshot Space -> Deploy UMA module through Zodiac app`

In other words, the user should be able to create an "optimistic DAO" without writing any code, through simple point-and-click interfaces. After that, users can handle governance actions through their familiar Snapshot interface.
